Fix issue #20044: Don't allow duplicate units in number with units interaction

JoaBrito opened the pull request:
## Overview
1. This PR fixes or fixes part of #20044.
2. This PR does the following: Detects and throws a message error when a unit is used more than once, either in the editor or in the interactions process.
3. (For bug-fixing PRs only) The original bug occurred because: The object factory never checked for it. So it was just creating and adjusting the condition.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

(I haven't done any tests cases since I'm not really aware what it should realistically cover. Could be a continuation of PR process)

## Proof that changes are correct

before:

https://github.com/oppia/oppia/assets/120871978/e109822a-034d-4f59-ae0b-40cd0af1e6a0

after:

https://github.com/oppia/oppia/assets/120871978/a260a266-474f-4d3a-900d-46698205cbc3

#### Proof of changes on desktop with slow/throttled network

https://github.com/oppia/oppia/assets/120871978/7c3d1d92-2b29-4bbe-881f-adf0c0c23f45

#### Proof of changes on mobile phone

https://github.com/oppia/oppia/assets/120871978/f1c1bb26-8354-463a-bc27-5a87f09aeb3f

#### Proof of changes in Arabic language
Not sure if this is Arabic, but it is a right to left language:

https://github.com/oppia/oppia/assets/120871978/6bbb1423-e2f0-45a3-bc5f-2273a454a14a

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

JoaBrito committed:
Fix Issue #20044 : Don't allow duplicate units in numbers during input interaction (#20044)

    * Created condition that checks for more than one unit use

    * Created a new error that appears when the new condition is verified

    * Created tests that show the new condition working

JoaBrito committed:
Fix #20044:Don't allow duplicate units in number with units interaction

JoaBrito committed:
Fix #20044:Don't allow duplicate units in number with units interaction

JoaBrito committed:
Fix #20044:Don't allow duplicate units in number with units interaction (error text size change)

JoaBrito committed:
Merge branch 'oppia:develop' into bugfix

JoaBrito committed:
Fix #20044:Don't allow duplicate units in number with units interaction (removed test cases)

JoaBrito committed:
Fix #20044:Don't allow duplicate units in number with units interaction (removed test cases)

JoaBrito commented:
@seanlip  PTAL. Since it's my first pull request can you please confirm that I've done everything alright?

seanlip commented:
@JoaBrito Many of your CI tests are failing. Could you please fix those? I think @AFZL210 can advise if you need help at any point, too.

AFZL210 commented:
@JoaBrito you can follow the guide given here to run frontend test locally https://github.com/oppia/oppia/wiki/Frontend-tests

You can even run frontend test for files which you have edited. just paste the test filename in ```combined-tests.spec.ts``` and once you have fixed test of your edited file undo the changes in ```combined-tests.spec.ts``` file and run all frontend test.

If you are stuck for too long while fixing test. feel free to ping me.

JoaBrito commented:
Still fixing some tests, but to improve test coverage what kind of tests should there be to fully test the new condition? I was thinking of using the cases that @seanlip used in the issue page, but if there is any other case I should had please tell me.

seanlip commented:
In general, you'll want to think about what behaviours the code supports and write tests to cover those behaviours (so that complete coverage happens as a side effect of that). See https://github.com/oppia/oppia/wiki/Backend-tests#testing-philosophy for an example -- the upshot is that you'd need to think through it.

JoaBrito commented:
Hey. I checked all the tests, ran prettier on my code changes to ensure code style and added some test cases to the files I changed. However when I tried to push the changes the following error appears. Is it safe to do the git command in the end of it? Wouldn't it change the git origin from the fork to the oppia repository?

![image](https://github.com/oppia/oppia/assets/120871978/be352a4f-d7d8-43d5-bc2e-61abadb47dda)

AFZL210 commented:
> Hey. I checked all the tests, ran prettier on my code changes to ensure code style and added some test cases to the files I changed. However when I tried to push the changes the following error appears. Is it safe to do the git command in the end of it? Wouldn't it change the git origin from the fork to the oppia repository?
> 
> ![image](https://private-user-images.githubusercontent.com/120871978/335841420-14f7affa-a092-4efb-88f7-104202469c7d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTczMzQ1OTEsIm5iZiI6MTcxNzMzNDI5MSwicGF0aCI6Ii8xMjA4NzE5NzgvMzM1ODQxNDIwLTE0ZjdhZmZhLWEwOTItNGVmYi04OGY3LTEwNDIwMjQ2OWM3ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwMlQxMzE4MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NTE3NWFhNjk0Yjg4OTNhMjgxNGNhMTYyZTJmYzJlMWZjMGFhNzBlODQ5OWFlZDQ2M2E0MjBiMmYzZWNmZTI0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.sfeJgTZLeCN7e1Plo1ROg6Ot-i3f2rUsw-otaUbQ0sU)

you need to set the upstream which is the main oppia codebase.

Run the commands given in the error.

```
git remote set-url upstream https://github.com/oppia/oppia.git
git remote add upstream https://github.com/oppia/oppia.git
```

JoaBrito committed:
Fix #20044:Don't allow duplicate units in number with units interaction (fixed failed checks and added test cases to modified files)

JoaBrito commented:
I just pushed the changes I was talking about yesterday.

Thanks @AFZL210, you're message helped a lot.

AFZL210 commented:
@JoaBrito Nice. Looks like all tests have passed.

Once you are satisfied with the changes you have made and want to get it reviewed, follow these steps:

1. Look at the names in the reviewers section (top right).
2. Click on the assignees setting box and search for the same usernames present in the reviewers section, or you can comment with the following message.

```@ReviewerUsername PTAL.```

JoaBrito commented:
@Nik-09 PTAL.

Nik-09 reviewed (changes requested):
Hi @JoaBrito, there are some units in which a physical quantity may repeat eg: 
[(Kg * m) / (s * s)]. I understand this can be written as [(Kg*m) / (s^2)]. 
Can you please provide a video, regarding what happens in these cases?

Thank you

rahat2134 committed:
Merge branch 'develop' into bugfix

JoaBrito commented:
Hi @Nik-09.  Currently, as you can see on the following video, there is no difference between those kinds of cases and a normal one like [km * km]. An error will always be showed if any unit is already in use. Should I make it so that your example as an exception? And if so, are there any other units/cases that should also follow that exception? 

https://github.com/oppia/oppia/assets/120871978/3826df38-e6d4-471a-9b1d-2c01eb110a29

Nik-09 commented:
> Hi @Nik-09. Currently, as you can see on the following video, there's no difference between those kinds of cases and a normal one like [km * km]. An error will always be shown if a unit is already in use. Should I make it so that your example works as an exception? And if so, are there any other units/cases that should also follow that exception?
> 
>  simplescreenrecorder-2024-06-12_03.39.25.mp4

Hi @JoaBrito this is particularly a product decision I think @seanlip will be able to help.

A summary: As per this PR, if a unit is repeated in an answer, an error message will be shown. But there are certain physical quantities in which units are expected to be duplicated e.g. Force in which unit is (Kg * m) / (sec * sec). Making the changes through this PR we will restrict learner to enter (Kg*m)/(sec^2).

@JoaBrito there might be other physical quantities like acceleration, where the units will repeat.

@seanlip can you please share your thoughts re this.
Thank you

seanlip commented:
@Nik-09 @JoaBrito I actually think that (kg*m)/(sec^2) is the right way to enter this, so the behaviour described in this PR seems correct to me. Have you actually seen (kg * m) / (sec * sec) used anywhere in practice?

JoaBrito commented:
Hi @Nik-09 @seanlip! Seems like a simple change, however I think that a bad error message could be a bit ambiguous to the user. Can I have confirmation on what the error message should look like?

Nik-09 commented:
Sorry for the delay in the review, I will do it tomorrow morning. Thank you.

seanlip commented:
> One suggestion might be, instead of just telling the user "your answer is formatted incorrectly", can we suggest to them the right formatting instead (in the case where we can determine what it is)?

How about: "Your answer has a repeated unit: {{unit_name}}. Try rewriting it as {{correct_version}}."

Nik-09 commented:
> @Nik-09 @JoaBrito I actually think that (kg*m)/(sec^2) is the right way to enter this, so the behaviour described in this PR seems correct to me. Have you actually seen (kg * m) / (sec * sec) used anywhere in practice?
> 
> One suggestion might be, instead of just telling the user "your answer is formatted incorrectly", can we suggest to them the right formatting instead (in the case where we can determine what it is)?

Hi @seanlip I have not encountered this in practice. I was thinking about different corner cases. Yes, we can suggest the right formatting to them. @JoaBrito any thoughts on this re implementation?
Thank you

Nik-09 commented on core/templates/domain/objects/UnitsObjectFactory.ts:
```diff
@@ -264,6 +264,37 @@ export class UnitsObjectFactory {
+    return instancesOfUnits;
+  }
+
+  duplicatedUnit(units: string): string {
```
Can we rename this to 
```suggestion
  getDuplicatedUnit(units: string): string {
```

Nik-09 commented on extensions/interactions/NumberWithUnits/directives/oppia-interactive-number-with-units.component.spec.ts:
```diff
@@ -271,4 +271,39 @@ describe('Number with units interaction component', () => {
     expect(component.componentSubscriptions.closed).toBeTrue();
   });
+
+  it('should show error when user submits answer contains dupplicated units', fakeAsync(() => {
```
'duplicated' wrong spelling

Nik-09 commented on assets/i18n/en.json:
```diff
@@ -716,6 +716,7 @@
     "I18N_INTERACTIONS_NUMBER_WITH_UNITS_INVALID_CURRENCY": "Please enter a valid currency (e.g., $5 or Rs 5)",
     "I18N_INTERACTIONS_NUMBER_WITH_UNITS_INVALID_CURRENCY_FORMAT": "Please write currency units at the beginning",
     "I18N_INTERACTIONS_NUMBER_WITH_UNITS_INVALID_UNIT_CHARS": "Please ensure that the unit only contains numbers, alphabets, (, ), *, ^, /, -",
+    "I18N_INTERACTIONS_NUMBER_WITH_UNITS_INVALID_DOUBLE_UNITS" : "The unit \"{duppedunit}\" is used more than once.",
```
Help me to understand what is duppedunit?

Nik-09 commented on extensions/interactions/NumberWithUnits/directives/oppia-interactive-number-with-units.component.ts:
```diff
@@ -75,9 +76,21 @@ export class InteractiveNumberWithUnitsComponent implements OnInit, OnDestroy {
             this.numberWithUnitsObjectFactory.fromRawInputString(newValue);
             this.errorMessageI18nKey = '';
             this.isValid = true;
+            this.problematicUnit = '';
```
Please rename this variable, "problematicUnit" is a very generic name.

Nik-09 commented on extensions/objects/templates/number-with-units-editor.component.ts:
```diff
@@ -37,6 +37,7 @@ export class NumberWithUnitsEditorComponent implements OnInit {
   @Input() value!: NumberWithUnitsAnswer | null;
   @Output() valueChanged = new EventEmitter();
   numberWithUnitsString!: string;
+  problematicUnit: string = '';
```
Ditto, please rename

Nik-09 reviewed (commented):
Hi @JoaBrito, added few comment PTAL

Nik-09 reviewed (changes requested):
Added review.

JoaBrito commented:
> Hi @seanlip I have not encountered this in practice. I was thinking about different corner cases. Yes, we can suggest the right formatting to them. @JoaBrito any thoughts on this re implementation? Thank you

Hi @Nik-09. Sorry for the delay on the re implementation. To do this I was thinking of when we detect a duplicated unit we would start adding/subtracting all the exponents based on what operation we see, ie. if ' ' or '*' is before a unit then we add the exponents or in the case of a '/', we would subtract them. Maybe I'm overthinking it or this tool is already implemented?

For the mean time, I'll start doing it based on this logic, but if there is a simpler way or a already existing tool that does this please let me know.

seanlip commented:
@JoaBrito There isn't a tool atm, but I think something along the lines of what you suggest could work. Perhaps might need to handle things like `kg/(m*s)` properly though, or raise an error due to the parens.

JoaBrito commented:
@seanlip I was thinking of that cases. Maybe throwing an error is too harsh, so i was trying to adapt the main ideia to make parens work aswell.

So as we have a multiplication we have a operation multiplier of 1 and if we see a division we have a multiplier of -1, ie  km*kg^-2 == km/kg^2 .This can work for parens, so if we see an parens open we will save the correspondent multiplier and then multiply every following operation with the new multiplier, until we see the parens close. This way kg/(m/s) == kg*(s/m).

I hope I made myself clear, sometimes it's a bit hard to explain this kind of ideias without clear examples to follow them with.

seanlip commented:
@JoaBrito It makes sense, but then you'll either need to generalize that to nested parens or disallow nested parens, I guess.

JoaBrito committed:
Merge branch 'develop' into bugfix

JoaBrito commented:
Hi@seanlip. I'm sorry that this has taken so long but I had a problem with my computer and I was unable to work on this issue. Since now I have everything back to normal and since it has been a while since we discussed this I would just like to confirm that the error message is in fact the last one that we discussed.
I already have the functionality of getting the corrected answer format from the duplicated input, and I only need to write some tests to improve coverage and do the requested changes that you told me to do.
Once again I'm sorry, I'll try to commit the already built in feature and all the other changes tomorrow.

seanlip commented:
Hi @JoaBrito, just following up on this -- sorry for the late response, I was out for the past few weeks and am only seeing this message now.

I do think that the error message mentioned [here](https://github.com/oppia/oppia/pull/20361#issuecomment-2182229118) seems like the right way to go. If you'd like to still push the code and wrap up this PR, that would be great, but if you can't, no worries. Thanks!
